openapi.yaml

```diff
@@ -471,6 +591,173 @@ paths:
           $ref: "#/components/responses/InternalServerError"
 components:
   schemas:
+    AddRelationArgs:
+      type: object
+      properties:
+        endpoints:
+          type: array
+          items:
+            type: string
+        via-cidrs:
+          type: array
+          items:
+            type: string
+      additionalProperties: false
+      required:
+        - endpoints
+    Application:
+      type: object
+      description: Information about a given application.
+      properties:
+        base:
+          $ref: '#/components/schemas/Base'
+        channel:
+          type: string
+        charm:
+          type: string
+        constraints:
+          $ref: '#/components/schemas/Constraints'
+        endpoint-bindings:
+          type: object
+          patternProperties:
+            .*:
+              type: string
+        exposed:
+          type: boolean
+        exposed-endpoints:
+          type: object
+          patternProperties:
+            .*:
+              $ref: '#/components/schemas/ExposedEndpoint'
+        life:
+          type: string
+        principal:
+          type: boolean
+        remote:
+          type: boolean
+        tag:
+          type: string
+      additionalProperties: false
+      required:
+        - tag
+        - principal
+        - exposed
+        - remote
+        - life
+    ApplicationSetCharm:
```

Comment:
I think this should probably be called something else (can't quite decide what, yet) and only include:

- `charm`: the name of the charm package (as per metadata.yaml)
- `channel`: a well defined CharmChannel as suggested above
- `origin`: a `CharmOrigin` object as it is here
- `url`: as defined here (I assume this points to a charmhub uri?)

Reply:
I changed its name to `ApplicationSubstitute` for now as a suggestion. It can always be changed and improved.

So currently in Juju the charm-url would have the charm name, unless it's a local charm, in which case it's something like `local:<path>`. But we really want to eventually get rid of the charm URL (we have been discussing that since Prague). As long as we have the origin (that has the source --charmhub/local--, base, track and the risk), we shouldn't have any trouble locating the charm. 

If the client is refreshing with a local charm, then it'll send the charm in the request body anyways, right? That part might need some modelling.

> and only include:

Are you suggesting we should not support things like resetting configs, e.g. `juju refresh foo --config config.yaml`, or uploading resources, e.g. `juju refresh foo --resource bar=/some/file.tgz` or `juju refresh foo --storage cache=ssd,10G`? Or are you just saying those can be added later once the main structure is established?

Reply:
Ok so happy with `ApplicationSubstitute` for now, though if we're including the `CharmOrigin` object as part of it then there are some fields that can be removed (`base`, `arch`, `track`, etc.) from the `ApplicationSubsitute`

You're right about `juju refresh foo --config ...` etc., we probably do want to allow for that.